### ISSUE TITLE
Exceptions should favor `raise` over `fail`

### ISSUE THREAD

**Issue by lsegal** [bbatsov/ruby-style-guide#512]
TL;DR `raise` is the correct verb to communicate raising an exception in Ruby. `fail` is not a reliable method, no better semantically, and simply, should not be favored over `raise`.

First off let me say that I'm all for a great community-driven style guide, and this is a great project to run! Thank you to the people who are making this possible! That said, I believe that to have a style guide that is reinforced by the community, it should represent the community. I currently feel as though the suggested use of `fail` does not represent the Ruby community. There are stats and reasons below.

To start off, I am trying to understand where this all came from-- i.e., the rationale behind favoring `raise` over `fail` in the style guide. I've never seen much `fail` usage in my time writing Ruby. The only relevant information I can find is from a [Jim Weirich email](http://devblog.avdi.org/2014/05/21/jim-weirich-on-exceptions/) published on Avdi Grimm's blog last year. From this reading, it seems as though the rationale for `fail` was taken verbatim from Jim's suggested preference rather than any deeper technical reasoning or look at real world usage.

Although I respect Jim greatly for his many contributions to the community, I would argue strongly against trying to shoehorn this personal preference into the community as a convention. Specifically, I would argue a few major points:

1) Regardless of how you create an exception, the verbiage that we use to communicate with Ruby developers is still "*raising an Exception*", not "failing a method" or anything of the sort. By that standard, any Ruby developer that sees `raise` should clearly understand what is happening. It does not seem as clear to me that `fail` would be universally understood, especially since there *are* other semantics at play here.

2) To the "other semantics" point: `fail` will *not work* on any class inheriting BasicObject:

```ruby
class Foo < BasicObject
  def foo; fail "bye" end
end

Foo.new.foo # => NoMethodError 
```

This exemplifies a very big technical flaw in pushing for `fail`. I'm not sure why we're recommending a convention that only works in "many cases" but "not all" cases. Incidentally, this style guide *makes no mention of this pitfall*. That said, I would not want to see the solution to this problem to just add a clause about BasicObject for a few reasons: (a) a developer may not know if their class actually inherits BasicObject (consider grand-child classes), and (b) recommending a syntax that is inherently unreliable never sounds like "good style" to me. In short, I would never suggest something that "works most of the time" as a "best practice". I don't think this style guide should, either.

3) Most importantly, though, if this project is reflecting the conventions of the community, it is currently "failing" (see what I did there!) with respect to `raise`/`fail`. Suggesting `fail` over `raise` does *not* reflect the Ruby community's existing conventions. Looking at a sampling of GitHub projects, even completely out of context, the term `raise` shows up more than 9 times more often than `fail`:

1. [`raise` in Ruby projects](https://github.com/search?l=ruby&q=raise&ref=searchresults&type=Code&utf8=%E2%9C%93) - ~9 million results
2. [`fail` in Ruby projects](https://github.com/search?l=ruby&q=fail&ref=searchresults&type=Code&utf8=%E2%9C%93) - ~1 million results

And given how often "fail" can be used to describe anything outside of an exception (think: any test case, [or any legitimate *redefinition of `def fail`, of which there are plenty*, unlike `raise`](https://github.com/search?l=ruby&p=6&q=%22def+fail%22&ref=searchresults&type=Code&utf8=%E2%9C%93)-- heck, just look at the first full page of results for fail and note that only 2/10 actually center around exception handling), it seems very clear that `raise` is the currently favored verb. Even GitHub's syntax highlighting on your own project page (and almost all syntax highlighting for Ruby) recognizes `raise` as a pseudo-keyword, while completely ignoring `fail`. Unless there is a strong technical reason, conventions should arise from community usage, not be prescribed upon them. In this case, there is no strong technical reason to favor `fail` over `raise`, only a personal preference-- and if "personal preference" is the argument for `fail`, then there is also far more "personal preference" on the side of `raise`.

I'm opening this issue to request the section about Exceptions be modified to suggest `raise` as standard practice for raising exceptions in Ruby. This is simpler for developers to follow (fewer verbs to be mindful of), has far less technical pitfalls, is more in line with Ruby terminology, and is far more conventional in terms of actual usage in the wild. I would further suggest that `fail` probably shouldn't even be mentioned in this style guide at all-- the argument being that the style guide exists to suggest "*best* practices", not "*second-best* practices"-- but I am a little flexible on this one.

FWIW, if there is agreement that this should be done, I'm more than happy to rewrite and submit a PR-- I just don't want to do it before I know it will be considered.

Thanks!

**Comment by scaryzet** [comment]
+1

Rarely seen `fail` while digging through the code of different gems.

`raise` is the community standard. `fail` must be deprecated.

Also ruby books teach you to use `raise` for raising exceptions.

**Comment by lee-jon** [comment]
+1

I've rarely seen `fail`, and although there is a similar debate on this in issue [#233](https://github.com/bbatsov/ruby-style-guide/issues/233) I don't think its a subjective one or one about building better exceptions.

**Comment by bestie** [comment]
+1

In addition to the great points by @lsegal

The idea that the exceptions I raise in my code today will never be rescued or that when rescuing an exception I have to go find the source and change fail -> raise is, to me, a little absurd.

**Comment by bbatsov** [comment]
I'm fine with changing this. Jim's suggestion seemed to reflect some of Ruby's design philosophy (and it encoded more meaning into the code), but I'm fond of simplicity and if it were up to me there would be 0 aliases in Ruby's core/standard library.

**Comment by lsegal** [comment]
:+1: on @lee-jon's PR.

**Comment by bbatsov** [comment]
The guide was just updated.

**Comment by csutter** [comment]
I'm a little bit late to the party, having just discovered this change trying to check something else in the guide, so apologies in advance for resurrecting this issue. Feel free to mutter under your breath about how annoying I am :wink: 

I use a `fail` most of the time unless re-raising previous exceptions, and like the urgency conferred by the term. I've long forgotten why I got into the habit in the first place, but I'm half tempted to blame an Avdi Grimm book after reading through the discussion in #233. I too had the impression that it was more true to Ruby's design philosophy, and feel like "X is more popular than Y" should not trump that. However I appreciate I'm only yet another single data point, and presumably only Matz can truly shed light on what the whole point of having two different verbs is.

What I can say though is that point (2) is definitely incorrect and shouldn't count as a reason to prefer `raise`. You are right that `BasicObject` doesn't support `fail`, however it doesn't support `raise` either! Both methods are part of `Kernel` which is not included in `BasicObject`. Also, they are both defined as the same C function `rb_f_raise` (see [Ruby's eval.c](https://github.com/ruby/ruby/blob/b6b1038c6bc487b0b8f7d1675d02a22a4ee8f4b8/eval.c#L1630)), so when they _are_ available their behaviour is absolutely identical in every way.

**Comment by perlun** [comment]
(Just FYI - Rubocop 0.37.0 includes this style change in its defaults now as well: https://github.com/bbatsov/rubocop/issues/2732)

**Comment by glennpratt** [comment]
I'm surprised this went in with so little discussion.

And as @csutter already pointed out:

```ruby
[1] pry(main)> class Foo < BasicObject
[1] pry(main)*   def foo; raise "bye" end
[1] pry(main)* end
=> :foo
[2] pry(main)> Foo.new.foo
NoMethodError: undefined method `raise' for #<Foo:0x007fef82adf178>
from (pry):6:in `foo'
```

**Comment by perlun** [comment]
> And as @csutter already pointed out:

Actually, it's the same with `fail`:

```
2.3.0 :001 > class Foo < BasicObject
2.3.0 :002?>   def foo; fail "bye" end
2.3.0 :003?>   end
 => :foo
2.3.0 :004 > Foo.new.foo
NoMethodError: undefined method `fail' for #<Foo:0x007fef83098cb8>
	from (irb):2:in `foo'
	from (irb):4
	from /Users/plundberg/.rvm/rubies/ruby-2.3.0/bin/irb:11:in `<main>'
```

**Comment by lsegal** [comment]
@csutter @glennpratt you're right that I completely blew it on point 2. I had done some testing and simply did not check my work properly there. Sorry about that! It is in fact true that raise/fail behave the same here.

That said, points 1 and 3 are equally valid, and probably even more meaningful than 2 is anyhow, since practically speaking few objects are actually inheriting BasicObject. I think the argument stands strong even without 2. Thank you for correcting me though.

**Comment by BBCMarkMcDonnell** [comment]
Just as an aside, this is my opinion/usage:

- Use `fail` to _signal_ exceptions
- Use `raise` to _rethrow_ exceptions

**Comment by jdickey** [comment]
@BBCMarkMcDonnell, are you saying that you've never rearranged/reused code such that in one use case an error was being signalled for the first time, and in another case *reusing the same code fragment* it was re-throwing an error? Do you really find that tenable? The Style Guide has flip-flopped on this before; I'd *far* rather see us choose one or the other and stick with it for either signalling or re-throwing.

**Comment by Integralist** [comment]
Personally I don't really mind. I just disable that particular check within Rubocop :-)

I merely mention how I use fail/raise as an aside to demonstrate how others use it. Whether it's taken into consideration or not, I leave that in the community's capable hands

**Comment by jdickey** [comment]
@Integralist, thanks for the clarification, and sorry for being harsh.

Note to self: never participate in issue discussions on what, in the freer world, would be sick days. 😷

**Comment by svnpenn** [comment]
FYI - the initial Ruby commit defined `fail` as an alias, so favoring `raise` makes sense:

http://github.com/ruby/ruby/blob/9c5b198/eval.c#L4214
